File: advertising_sdk/src/main/java/com/adadapted/android/sdk/core/ad/AdContent.java
package com.adadapted.android.sdk.core.ad;

import android.os.Parcel;
import android.os.Parcelable;

import com.adadapted.android.sdk.core.atl.AddToListContent;
import com.adadapted.android.sdk.core.atl.AddToListItem;
import com.adadapted.android.sdk.core.event.AppEventClient;

import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class AdContent implements AddToListContent, Parcelable {
    private static final String LOGTAG = AdContent.class.getName();

    public static final int ADD_TO_LIST = 0;

    private final Ad ad;
    private final int type;
    private final List<AddToListItem> items;

    private boolean handled;
    private final Lock lock = new ReentrantLock();

    protected AdContent(Parcel in) {
        ad = in.readParcelable(Ad.class.getClassLoader());
        type = in.readInt();
        items = in.createTypedArrayList(AddToListItem.CREATOR);
        handled = in.readByte() != 0;
    }

    public static final Creator<AdContent> CREATOR = new Creator<AdContent>() {
        @Override
        public AdContent createFromParcel(Parcel in) {
            return new AdContent(in);
        }

        @Override
        public AdContent[] newArray(int size) {
            return new AdContent[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeParcelable(ad, i);
        parcel.writeInt(type);
        parcel.writeTypedList(items);
        parcel.writeByte((byte) (handled ? 1 : 0));
    }

    private AdContent(final Ad ad,
                      final int type,
                      final List<AddToListItem> items) {
        this.ad = ad;
        this.type = type;
        this.items = items;

        handled = false;
    }

    public static AdContent createAddToListContent(final Ad ad) {
        if (ad.getPayload().size() == 0) {
            AppEventClient.trackError(
                "AD_PAYLOAD_IS_EMPTY",
                String.format(Locale.ENGLISH, "Ad %s has empty payload", ad.getId())
            );
        }

        return new AdContent(ad, ADD_TO_LIST, ad.getPayload());
    }

    public synchronized void acknowledge() {
        lock.lock();
        try {
            if(handled) {
                return;
            }

            handled = true;
            AdEventClient.trackInteraction(ad);
        }
        finally {
            lock.unlock();
        }
    }

    @Override
    public synchronized void itemAcknowledge(final AddToListItem item) {
        lock.lock();
        try {
            if (!handled) {
                handled = true;
                AdEventClient.trackInteraction(ad);
            }

            trackItem(item.getTitle());
        }
        finally {
            lock.unlock();
        }
    }

    private synchronized void trackItem(final String itemName) {
        final Map<String, String> params = new HashMap<>();
        params.put("ad_id", ad.getId());
        params.put("item_name", itemName);

        AppEventClient.trackSdkEvent("atl_added_to_list", params);
    }

    public synchronized void failed(final String message) {
        lock.lock();
        try {
            if(handled) {
                return;
            }

            handled = true;
            final Map<String, String> params = new HashMap<>();
            params.put("ad_id", ad.getId());
            AppEventClient.trackError("ATL_ADDED_TO_LIST_FAILED",
                    (message == null || message.isEmpty()) ? "Unknown Reason" : message,
                    params);
        }
        finally {
            lock.unlock();
        }
    }

    @Override
    public void itemFailed(final AddToListItem item, final String message) {
        lock.lock();
        try {
            handled = true;
            final Map<String, String> params = new HashMap<>();
            params.put("ad_id", ad.getId());
            params.put("item", item.getTitle());
            AppEventClient.trackError("ATL_ADDED_TO_LIST_ITEM_FAILED",
                    (message == null || message.isEmpty()) ? "Unknown Reason" : message,
                    params);
        }
        finally {
            lock.unlock();
        }
    }

    public String getZoneId() {
        return ad.getZoneId();
    }

    public int getType() {
        return type;
    }

    public List<AddToListItem> getItems() {
        return items;
    }

    public boolean hasItems() {
        return items.size() > 0;
    }

    public boolean hasNoItems() {
        return items.size() == 0;
    }

    public boolean isHandled() {
        return handled;
    }

    public String getSource() {
        return Sources.IN_APP;
    }

    @Override
    public String toString() {
        return "AdContent{" +
            "zone='" + ad.getZoneId() + '\'' +
            "items=" + items +
            '}';
    }
}

File: advertising_sdk/src/main/java/com/adadapted/android/sdk/core/ad/AdContent.kt
package com.adadapted.android.sdk.core.ad

import com.adadapted.android.sdk.config.EventStrings
import com.adadapted.android.sdk.core.atl.AddToListContent
import com.adadapted.android.sdk.core.atl.AddToListItem
import kotlin.collections.HashMap

class AdContent private constructor(
    private val ad: Ad,
    private val items: List<AddToListItem>,
    private val eventClient: EventClient = EventClient
) : AddToListContent {
    private var isHandled: Boolean

    override fun acknowledge() {
        if (isHandled) {
            return
        }
        isHandled = true
        eventClient.trackInteraction(ad)
    }

    @Synchronized
    override fun itemAcknowledge(item: AddToListItem) {
        if (!isHandled) {
            isHandled = true
            eventClient.trackInteraction(ad)
        }
        trackItem(item.title)
    }

    @Synchronized
    private fun trackItem(itemName: String) {
        val params: MutableMap<String, String> = HashMap()
        params[AD_ID] = ad.id
        params[ITEM_NAME] = itemName
        eventClient.trackSdkEvent(EventStrings.ATL_ITEM_ADDED_TO_LIST, params)
    }

    @Synchronized
    override fun failed(message: String) {
        if (isHandled) {
            return
        }
        isHandled = true
        val params: MutableMap<String, String> = HashMap()
        params[AD_ID] = ad.id
        eventClient.trackSdkError(
            EventStrings.ATL_ADDED_TO_LIST_FAILED,
            message.ifEmpty { UNKNOWN_REASON },
            params
        )
    }

    override fun itemFailed(item: AddToListItem, message: String) {
        isHandled = true
        val params: MutableMap<String, String> = HashMap()
        params[AD_ID] = ad.id
        params[ITEM] = item.title
        eventClient.trackSdkError(
            EventStrings.ATL_ADDED_TO_LIST_ITEM_FAILED,
            message.ifEmpty { UNKNOWN_REASON },
            params
        )
    }

    val zoneId: String
        get() = ad.zoneId

    override fun getItems(): List<AddToListItem> {
        return items
    }

    override fun hasNoItems(): Boolean {
        return items.isEmpty()
    }

    override fun getSource(): String {
        return AddToListContent.Sources.IN_APP
    }

    companion object {
        private const val AD_ID = "ad_id"
        private const val ITEM_NAME = "item_name"
        private const val ITEM = "item"
        private const val UNKNOWN_REASON = "Unknown Reason"

        fun createAddToListContent(ad: Ad): AdContent {
            return AdContent(ad, ad.payload.detailedListItems)
        }
    }

    init {
        if (ad.payload.detailedListItems.isEmpty()) {
            eventClient.trackSdkError(
                EventStrings.AD_PAYLOAD_IS_EMPTY,
                "Ad ${ad.id} has empty payload"
            )
        }
        isHandled = false
    }
}